### PR TITLE
Draft of preprocess post

### PR DESCRIPTION
**Pull request by Deraen**
https://dev.clojure.org/jira/browse/CLJS-2143

### CODE REVIEW

**Comment by Deraen** [comment]
One thing I noticed, we should probably decide on using either preprocess or transform term, not both.

I think the only places using "transform" are the JS modules guide and the multimethod name.

**Comment by swannodette** [comment]
As with @anmonteiro this post could do with a more contextual preparation for the reader. First we should start with rationale - why would a very different compile to JavaScript language care about this? 

Interop with the wider JS ecosystem! ClojureScript unlike some other compile-to-JS solutions is pushing harder, and harder to integrate with mainstream JavaScript practice and libraries. While Closure gives us AMD, CommonJS, ES6 module transformation, it doesn't handle other kinds of popular transforms - for example Babel transforms - like JSX. By allowing this type of transform we're furthering our reach even more.

**Comment by mike-thompson-day8** [comment]
Could I suggest the following structure  (a bit sheepishly, because I'm not the one doing the hard work on this) ...

  - Introduction or Abstract or TL;DR  (terse. Two sentences)
  - Motivation (Why? What problem is solved or what benefit arises. aka Vitamin or pain killer?)
  - What has changed and how to use it   (multipart explanation? Specifics. Code samples etc)

Further, optional sections (often unnecessary):
  - Q&A
  - Background
  - Forces   (acting on this decision, including constraints)
  - Examples
  - Previous Work
  - Rejected Ideas
  - Future Directions

I, personally, find these headings often help to inspire me towards clearer, more forceful technical explanations.   But, yeah, I'm not directly involved in this, so perhaps this is just unhelpful noise.

**Comment by mfikes** [comment]
One thing that caught me off guard: I didn't realize that the preprocess symbol referred to a symbol in a Clojure namespace. I spent a few seconds thinking about it how this would be loading ClojureScript code, and then realized, "oh, right, it has to be a symbol referring to _Clojure_ code". I suppose that an improvement could be to indicate that the transformation namespace can be arbitrary Clojure code that you have on the classpath.

File: content/news/2017-07-xx-js-preprocessing-improvements.adoc [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+2017-07-09
+:jbake-type: post
+
+ifdef::env-github,env-browser[:outfilesuffix: .adoc]
+
+This is the fourth post in the link:/news/2017-07-07-sneak-preview[Sneak Preview] series.
+
+// Introduction
+
+Closure compiler can process AMD, CommonJS and ES6 modules giving a good support for
```
Comment (swannodette) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: drop `a`

File: content/news/2017-07-xx-js-preprocessing-improvements.adoc [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+link:/news/2017-07-12-clojurescript-is-not-an-island-integrating-node-modules[leveraging Node ecosystem].
+// not sure this is worth mention, as I can't find good source for what is really supported
+// footnote:[Closure also supports TypeScript, but this has not been tested with Cljs yet.]
```
Comment (swannodette) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: This is interesting! Not related to this post but do you have a link?

Comment (Deraen) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: https://github.com/google/closure-compiler/search?utf8=%E2%9C%93&q=typescript&type=

https://github.com/google/closure-compiler/wiki/Releases
> February 2, 2015 (v20150126)
> Initial work on supporting TypeScript-style type annotations, with the --language_in=ES6_TYPED flag.

There doesn't seem to be many commits lately mentioning typescript, so perhaps the support is just for "TypeScript-style type annotations"

**Comment by Deraen** [comment]
- I fixed some typos I noticed.
- I added mention how using Nashorn allows using existing JS code to transform code.
- I reworked the solution part to describe some alternative solutions and how the final solution was selected

File: content/news/2017-07-xx-js-preprocessing-improvements.adoc [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+However, there are cases which Closure doesn't support directly, like JSX.
```
Comment (Deraen) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: Done

File: content/news/2017-07-xx-js-preprocessing-improvements.adoc [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+footnote:[There is https://github.com/mihaip/react-closure-compiler[third-party compiler pass] for Closure to support JSX]
+Support for link:/guides/javascript-modules#babel-transforms[transforming JS code]
```
Comment (Deraen) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: I already wrote about Babel later in the post.

File: content/news/2017-07-xx-js-preprocessing-improvements.adoc [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+using ClojureScript compiler has been around since
+https://github.com/clojure/clojurescript/blob/master/changes.md#1748[version 1.7.48].
+This feature allows transformation of the JS code before the code is passed into Closure compiler.
+The next ClojureScript version will improve this feature.
```
Comment (Deraen) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: Done

File: content/news/2017-07-xx-js-preprocessing-improvements.adoc [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+=== Motivation
+
+Most of the JavaScript libraries using JSX or other https://babeljs.io/[Babel] extensions
+are packaged with already processed code, so they can be used in ClojureScript like
+vanilla JS libraries. However, there are cases where one wants to include such code in
+the project directly. For example, when converting a project from JS to Cljs it cab be useful to
```
Comment (Deraen) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: Done

File: content/news/2017-07-xx-js-preprocessing-improvements.adoc [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+use the existing code, instead of rewriting everything in one go. Another case is
+when UI designers who want to work with JSX.
+footnote:[Though in my experience, UI designers are able to write Hiccup as easily.]
+
+// Technical Motivation
+
+== Preprocess option
+
+The option is enabled by providing `:preprocess` option to a link:/reference/compiler-options#foreign-libs[foreign-lib map].
```
Comment (Deraen) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: Done

File: content/news/2017-07-xx-js-preprocessing-improvements.adoc [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+The value is a keyword and is used as dispatch value for `cljs.closure/js-transforms`
+multimethod.  Implementing a transformation requires writing a new method for that multimethod.
+The namespace that proivdes this method has to be loaded before ClojureScript
+compiler is run. One problem has been that there has not been support in popular build tools
+for this feature.
+
+Due to how both Lein and Boot isolate the build tooling to their own classpaths, loading the namespace
```
Comment (Deraen) [inline content/news/2017-07-xx-js-preprocessing-improvements.adoc]: Done

**Pull request merged**